test/e2e/specs/editor/various/write-design-mode.spec.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * WordPress dependencies
+ */
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+
+test.describe( 'Write/Design mode', () => {
+	test.beforeEach( async ( { admin, editor } ) => {
+		await admin.createNewPost();
+		await expect(
+			editor.canvas.getByRole( 'textbox', { name: 'Add title' } )
+		).toBeFocused();
+	} );
+
+	test.afterAll( async ( { requestUtils } ) => {
+		await requestUtils.deleteAllPosts();
+	} );
+
+	test( 'Should prevent selecting intermediary blocks', async ( {
+		editor,
+		page,
+	} ) => {
+		// Insert a section with a nested block and an editable block.
+		await editor.insertBlock( {
+			name: 'core/group',
+			attributes: {
+				style: {
+					spacing: {
+						padding: '20px',
+					},
+					color: {
+						background: 'darkgray',
+					},
+				},
+			},
+			innerBlocks: [
+				{
+					name: 'core/group',
+					attributes: {
+						style: {
+							spacing: {
+								padding: '20px',
+							},
+							color: {
+								background: 'lightgray',
+							},
+						},
+					},
+					innerBlocks: [
+						{
+							name: 'core/paragraph',
+							attributes: {
+								content: 'Something',
+							},
+						},
+					],
+				},
+			],
+		} );
+
+		// Switch to write mode.
+		await editor.switchEditorTool( 'Write' );
+
+		const sectionBlock = editor.canvas
+			.getByRole( 'document', {
+				name: 'Block: Group',
+			} )
+			.nth( 0 );
+		const sectionClientId = await sectionBlock.getAttribute( 'data-block' );
+		const nestedGroupBlock = sectionBlock.getByRole( 'document', {
+			name: 'Block: Group',
+		} );
+		const paragraph = nestedGroupBlock.getByRole( 'document', {
+			name: 'Block: Paragraph',
+		} );
+		const paragraphClientId = await paragraph.getAttribute( 'data-block' );
+
+		// We should not be able to select the intermediary group block.
+		// if we try to click on it (the padding area)
+		// The selection should land on the top level block.
+		const nestedGroupPosition = await nestedGroupBlock.boundingBox();
+		await page.mouse.click(
+			nestedGroupPosition.x + 5,
+			nestedGroupPosition.y + 5
+		);
+
+		const getSelectedBlock = async () =>
+			await page.evaluate( () =>
+				window.wp.data
+					.select( 'core/block-editor' )
+					.getSelectedBlockClientId()
+			);
+
+		expect( await getSelectedBlock() ).toEqual( sectionClientId );
+
+		// We should be able to select the paragraph block and write in it.
+		await paragraph.click();
+		await page.keyboard.type( ' something' );
+		expect( await getSelectedBlock() ).toEqual( paragraphClientId );
+		expect( await paragraph.innerHTML() ).toEqual( 'Something something' );
+
+		// Check that the inspector still shows the group block with the content panel.
+		await editor.openDocumentSettingsSidebar();
+		const editorSettings = page.getByRole( 'region', {
+			name: 'Editor settings',
+		} );
+		await expect(
+			editorSettings.locator( '.block-editor-block-card__title' )
```

Comment:
Using class name selectors is not recommended. Could we instead select it by `getByRole` and optionally with a `level` field here?

Reply:
I know but for me, this is not easy to pinpoint and avoid random failures when we add headings... I prefer the selector here.

Reply:
I think there's already a heading in place? What do you mean by pinpointing random failures? I'd appreciate if we could follow the established [best practices](https://github.com/WordPress/gutenberg/tree/trunk/docs/contributors/code/e2e#use-accessible-selectors) whenever possible and add comments if we could not. Thanks! 🙏

Reply:
The `h2` is not enough to say that I'm targeting the "title" of the inspector panel. There could be more h2 in the inspector panel, in fact there are already. (content).

We could theoretically say `h2 > eq( 0 )` but that's not very stable, any random design change can break that. (moving titles around).

So for me, in this case, I prefer a non ambiguous assertion rather than a generic one. 

I agree on following best practices, but we need to understand what's important and why they exist. 

I'm happy to update If there's a less ambiguous approach using that doesn't use selectors.

Reply:
A common workaround practice throughout the codebase is using `expect( getByRole('heading', {name: 'Group', level: 2}) ).toBeVisible()`.

Reply:
The problem is that for end users there's no way to connect `.block-editor-block-card__title` to the block card title, they are not even aware of the class name. A simple way to think of this is from the perspective of screen readers. What screen readers read is what we test here. If we want to bring more context to the headings we should make that information accessible to them too.

Reply:
That makes a lot of sense I agree. How can we do that?

Reply:
Let's stick with the current practice of `getByRole` and figure out the accessibility concern in a new issue?

Reply:
For me, when this type of issue arrises it's an indicator that the a11y of the code in question could be improved.

Riad is correct the `getByRole` is (in this particular instance) ambiguous and could be fragile.

Kai is right that we should strive to interact with perceivable elements rather than rely on "developer APIs" such as CSS classnames. 

Can we improve the markup here to allow the selector to be less ambiguous? For me it feels like the name of the block is the title for the _entire_ sidebar so it should be a higher heading level than those used for the panels below. That would indicate a hierarchy, where the block controls themselves are "owned" by the "block".

We could also consider using [`<hgroup>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/hgroup)

> The <hgroup> element represents a heading and related content. 

Or even `<header>` if we introduced a sectioning level element.

Reply:
I feel like until we fix a potential a11y thing here, we should stick with the class for the test as it's more solid.

Reply:
On the a11y approach, I think it's less about heading levels, and more about having an explicit aria-label or something like `Selected block type is "something"` but I may be wrong there and welcome alternatives.

Reply:
I'd echo Riad about until fixing the a11y issue, we should stick with that classname for now and unblock this PR. Let's create an issue for this.

Reply:
No need to block this PR. But we should raise a followup Issue [as suggested](https://github.com/WordPress/gutenberg/pull/65819#discussion_r1787374700) with the appropriate labels to get Accessibility feedback.

Reply:
IMO, `getByRole` is always preferred, in cases when it doesn't apply, we should still fallback to alternatives like `getByTestId` than class name selectors. I don't understand the argument of using class name selectors being more "solid". TBH, this has happened too many times that this is becoming frustrating to me. Could we at least add a comment above indicating that this is an edge case so that people won't accidentally follow the practice? I don't think it should block this PR, but I think this type of issue is still important, especially given that it just flagged a potential accessibility concern here by trying to avoid the class name selectors.

Reply:
I'll add a comment. I'm not sure we used test ids before.